lib/api/api.go

```diff
@@ -1617,6 +1631,16 @@ func toJsonFileInfoSlice(fs []db.FileInfoTruncated) []jsonFileInfoTrunc {
 	return res
 }
 
+func toJsonPendingDeviceSlice(devices map[protocol.DeviceID]db.ObservedDevice) []jsonPendingDevice {
+	res := make([]jsonPendingDevice, len(devices))
+	i := 0
+	for id, meta := range devices {
+		res[i] = jsonPendingDevice{id.String(), meta}
+		i++
```

Comment:
You can make the slice with 0 length and appriopriate length and then do `res = append(res, jsonPendingDevice{...}` without needing `i`.

Also I don't quite understand why you need to manually handle the json serialisation at all: Couldn't you just add the device ID to db.ObservedDevice and use the default serialisation (as in proto-generated json annotations)?

Reply:
IIUC using `append()` will trigger a reallocation every time if the capacity is too small (`cap(res) = 0` as you'd suggest).  That's what I try to avoid by getting the map langth.

Storing the device ID in `ObservedDevice` would be duplicated. It makes sense in the XML config, but in the DB it already needs to be part of the key.

Reply:
Plus the serialization of slices is not something that happens automatically.  I pretty much copied the `toJsonFileInfoSlice()` function above, just using a map for the above reason.

Reply:
Sorry, mis-typed above: If you generate the slice with the right *capacity*, append will not allocate.

There's not general problem with slices for json as far as I know. The file info slice handling is due to custom serialisation of file info members. I thought that wasn't necessary for your types, but I didn't consider the time for serialisation, that does indeed need special handling.

And the duplication argument also applies here in the json map. I'd either add it on both sides or neither.

Reply:
JSON handles slices just fine.  What's happening here is a conversion / flattening of a map 

    {device ID -> {name, address, time}, ...}

to a flat JSON array of JSON objects

    [{deviceID, name, time, address}, ...]

which I think will be easier to consume in the GUI.  It makes more sense when looking at the related, planned [candidate devices API](https://deploy-preview-499--syncthing-docs.netlify.com/rest/system-candidates-get.html).  In that case, all details except the device ID may be missing in case it refers to an already known remote device.  I wanted to keep the APIs consistent with regard to where the device ID lives.

Reply:
```suggestion
	res := make([]jsonPendingDevice, 0, len(devices))
	for id, meta := range devices {
		res = append(res, jsonPendingDevice{id.String(), meta})
```

Reply:
Okay, now I understand what you meant with supplying a non-zero capacity but zero length. Wasn't quite familiar with the three-argument form of `make()`.